scanner/engine/save_worker.cpp

```diff
@@ -72,68 +75,133 @@ void* save_thread(void* arg) {
       WriteFile* output_file = nullptr;
       BACKOFF_FAIL(storage->make_write_file(output_path, output_file));
 
-      if (work_entry.columns[out_idx].rows.size() != num_rows) {
-        LOG(FATAL) << "Output layer's row vector has wrong length";
+      if (work_entry.columns[out_idx].size() != num_elements) {
+        LOG(FATAL) << "Output layer's element vector has wrong length";
       }
 
-      if (!work_entry.column_handles[out_idx].is_same_address_space(
-              CPU_DEVICE)) {
-        std::vector<u8*> dest_buffers, src_buffers;
-        std::vector<size_t> sizes;
-        size_t total_size = 0;
-        for (i32 f = 0; f < num_rows; ++f) {
-          Row& row = work_entry.columns[out_idx].rows[f];
-          total_size += row.size;
-        }
-
-        if (num_rows > 0) {
-          u8* output_block = new_block_buffer(CPU_DEVICE, total_size, num_rows);
-          for (i32 f = 0; f < num_rows; ++f) {
-            Row& row = work_entry.columns[out_idx].rows[f];
-            size_t size = row.size;
-            u8* src_buffer = row.buffer;
-            u8* dest_buffer = output_block;
-
-            dest_buffers.push_back(dest_buffer);
-            src_buffers.push_back(src_buffer);
-            sizes.push_back(size);
+      // Ensure the data is on the CPU
+      move_if_different_address_space(args.profiler,
+                                      work_entry.column_handles[out_idx],
+                                      CPU_DEVICE, work_entry.columns[out_idx]);
 
-            output_block += size;
+      bool compressed = work_entry.compressed[out_idx];
+      // If this is a video...
+      i64 size_written = 0;
+      if (work_entry.column_types[out_idx] == ColumnType::Video) {
+        // Read frame info column
+        assert(work_entry.columns[out_idx].size() > 0);
+        FrameInfo frame_info = work_entry.frame_sizes[video_col_idx];
+
+        // Create index column
+        VideoMetadata video_meta;
+        proto::VideoDescriptor& video_descriptor = video_meta.get_descriptor();
+        video_descriptor.set_table_id(io_item.table_id());
+        video_descriptor.set_column_id(out_idx);
+        video_descriptor.set_item_id(io_item.item_id());
+
+        video_descriptor.set_width(frame_info.width());
+        video_descriptor.set_height(frame_info.height());
+        video_descriptor.set_channels(frame_info.channels());
+        video_descriptor.set_frame_type(frame_info.type);
+
+        video_descriptor.set_time_base_num(1);
+        video_descriptor.set_time_base_denom(25);
+
+        if (compressed &&
+            frame_info.type == FrameType::U8 &&
+            frame_info.channels() == 3) {
+          H264ByteStreamIndexCreator index_creator(output_file);
```

Comment:
Since the encoder is in the save worker, this now means that if you want to use all your cores for encoding, then you need # save workers = # cores, right? This seems a little weird since it conflates the I/O with the compute?